rllib/algorithms/algorithm_config.py

```diff
@@ -2601,10 +2577,54 @@ def get_default_rl_module_class(self) -> Union[Type["RLModule"], str]:
 
         Returns:
             The RLModule class to use for this algorithm either as a class type or as
-            a string (e.g. x.y.z).
+            a string (e.g. ray.rllib.core.rl_trainer.testing.torch.BCModule).
+        """
+        raise NotImplementedError
+
+    def get_default_rl_trainer_class(self) -> Union[Type["RLTrainer"], str]:
+        """Returns the RLTrainer class to use for this algorithm.
+
+        Override this method in the sub-class to return the RLTrainer class type given
+        the input framework.
+
+        Returns:
+            The RLTrainer class to use for this algorithm either as a class type or as
+            a string (e.g. ray.rllib.core.rl_trainer.testing.torch.BCTrainer).
         """
         raise NotImplementedError
 
+    def get_trainer_runner_config(
+        self, observation_space: Space, action_space: Space, validate: bool = False
+    ) -> TrainerRunnerConfig:
+
+        if not self._is_frozen:
+            raise ValueError(
+                "Cannot call `get_trainer_runner_config()` on an unfrozen "
+                "AlgorithmConfig! Please call `freeze()` first."
+            )
+
+        config = (
+            TrainerRunnerConfig()
+            .module(
+                module_class=self.rl_module_class,
+                observation_space=observation_space,
+                action_space=action_space,
+                model_config=self.model,
+            )
+            .trainer(
+                trainer_class=self.rl_trainer_class,
+                eager_tracing=self.eager_tracing,
+                # TODO (Kourosh): optimizer config can now be more complicated.
+                optimizer_config={"lr": self.lr},
+            )
+            .resources(num_gpus=self.num_gpus, fake_gpus=self._fake_gpus)
+        )
+
+        if validate:
+            config.validate()
```

Comment:
why do you need this? validate() should be called by the user of AlgorithmConfig. I don't think we want the validate param here.
if validate() is not called, you are not guaranteed to have self.rl_trainer_class, so things may not work.
we should maybe have a flag is_validated instead, and err on that here?

Reply:
so in my opinion, we should only have 1 validate(), which is the top level validate().
it may be cleaner to always do:
```
config = PPOAlgorithmConfig(...).rl_module(...)
config.validate()

trainer_runner = config.get_trainer_runner_config().build()
```
like you basically finish all the configurations together in the top level validate() call. then simply get the trainer_runner_config. get_trainer_runner_config() in this case will be a read-only getter.
just a suggestion.

Reply:
I like this but implementation wise, you cannot validate the runner config unless it's created via 'get_trainer_runner_config'. Because the logic for its validation is located in its instance. So in practice how should i do it?

Reply:
wait, I don't quite get it.
can your ``trainer_runner_config`` has a ``validate()`` function that only gets called by ``AlgorithmConfig.validate()``?

Reply:
where would I build the trainer_runner_config object to call its `validate()` ? inside validate of `AlgorithmConfig`?